### PR TITLE
Support echidna invariant.

### PR DESCRIPTION
**Pull request by Cloud10240**

### CODE REVIEW

File: src/evm/oracles/echidna.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            let echidna_txn = Bytes::from(echidna_func.clone());
+            let addr = if self.address.is_zero() {
+                ctx.input.contract
+            } else {
+                self.address
+            };
+
+            let data = vec![(addr, echidna_txn)];
+
+            let res = ctx.call_post_batch(&data);
```
Comment (shouc) [inline src/evm/oracles/echidna.rs:66]: conduct batch call here to save time (i.e., create all the echidna calls at once and call `call_post_batch`)

File: src/evm/oracles/echidna.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                .all(|x| x);
+
+            if (res_bool) {
+                unsafe {
+                    ORACLE_OUTPUT = format!(
+                        "[echidna] echidna invariant violated {:?}",
+                        ctx.input.contract
+                    )
+                }
+                return vec![FUNCTION_BUG_IDX];
```
Comment (shouc) [inline src/evm/oracles/echidna.rs:84]: shall be `ECHIDNA_BUG_IDX`

**Comment by rappie** [comment]
This looks really interesting! I use Echidna a lot.

Would it be possible to give a bit of information on how to set this up? I'm happy to do some testing :smile:

**Comment by shouc** [comment]
@Cloud10240 pls add docs when you r free

**Comment by shouc** [comment]
hey @rappie thx for interests. It currently does not support echidna way of deployment. ItyFuzz just finds all violations (`echidna_*` returns false) by default.

**Comment by rappie** [comment]
Which way of deployment do you mean? Using `etheno`?

Will it work if I deploy everything in the constructor?

**Comment by shouc** [comment]
Yes, `etheno`doesnt work

For constructor deployment, likely not if you are using `CREATE` since no ABI is provided, but we are working on supporting it soon (providing extra args to specify ABI). Right now, the best way is just to follow https://github.com/fuzzland/ityfuzz#constructor-arguments

**Pull request merged**